\documentclass[article,nojss]{jss}


<<package,include=FALSE,echo=FALSE>>=
options(keep.source = TRUE, width = 60)
packageInfo <- packageDescription("DNAPlotR")
library(DNAPlotR)
packageKeywords<-"visualization, alignment, display, genome, DNA, sequence, multiple sequence alignment, reference sequence"
@
%\VignetteIndexEntry{\Sexpr{packageInfo$Package} examples}
%\VignetteDepends{}
%\VignetteKeywords{\Sexpr{packageKeywords}}
%\VignettePackage{\Sexpr{packageInfo$Package}}

\title{\pkg{\Sexpr{packageInfo$Package}} package example (version \Sexpr{packageInfo$Version})}
\author{\Sexpr{packageInfo$Author}}
\Plainauthor{\Sexpr{packageInfo$Author}}

\Address{
  Github: \url{http://github.com/sherrillmix/dnaplotr}
}

\Keywords{\Sexpr{packageKeywords}}

\Abstract{
  This is a collection of examples of usage for the \pkg{\Sexpr{packageInfo$Package}} package.
}

\begin{document}
\SweaveOpts{engine=R,eps=FALSE}

\section{General description}

\pkg{\Sexpr{packageInfo$Package}} allows rapid visual assessment of many DNA, RNA or amino acid sequences by plotting each sequence as a row of colors with each color representing a base/amino acid. A simple example of using the \pkg{DNAPlotR} package in R (Figure \ref{figPlotDNA}):
<<plotDNA, echo=TRUE, eval=FALSE>>=
  set.seed(1234)
  fakeSeqs<-createFakeDNA(5000,500)
  refSeq<-fakeSeqs[1]
  fakeSeqs<-fakeSeqs[-1]
  species<-sprintf('Species %s',sub(' [0-9]+$','',names(fakeSeqs)))
  par(mar=c(3.5,4.4,.5,7),mgp=c(2.5,1,0))
  plotDNA(fakeSeqs,groups=species)
@

\begin{figure}[htbp]
\begin{center}
<<showPlotDNA, fig=TRUE, height=4.5, width=8, echo=FALSE>>=
<<plotDNA>>
@
\label{figPlotDNA}
\caption{An example of a comparison of many sequences aligned against a reference}
\end{center}
\end{figure}


The package is also useful to visualize high-throughput, short read data in a given region (Figure \ref{figShortReads}):
<<plotShortReads, echo=TRUE, eval=FALSE>>=
  set.seed(1234)
  seqLength<-1000
  fakeSeqs<-createFakeDNA(5000,seqLength,pGap=0)
  refSeq<-fakeSeqs[1]
  fakeSeqs<-fakeSeqs[-1]
  potentialStarts<-1:(seqLength-99)
  #leave a gap in sequences
  potentialStarts<-potentialStarts[potentialStarts>550|potentialStarts<400]
  startCoords<-sort(sample(potentialStarts,5000,TRUE))
  endCoords<-startCoords+99
  dummy<-paste(rep('-',seqLength),collapse='')
  substring(fakeSeqs,1,startCoords)<-substring(dummy,1,startCoords)
  substring(fakeSeqs,endCoords+1,seqLength)<-substring(dummy,endCoords+1,seqLength)
  fakeSeqs<-replaceOuterGaps(fakeSeqs)
  par(mar=c(3.5,4.4,.5,1),mgp=c(2.5,1,0))
  plotDNA(fakeSeqs)
@

\begin{figure}[htbp]
\begin{center}
<<showShortReads, fig=TRUE, height=4.5, width=8, echo=FALSE>>=
<<plotShortReads>>
@
\caption{An example plot of short reads aligned against a reference sequence. The gap in coverage might indicate an indel, mapping problem or incorrect reference.}
\label{figShortReads}
\end{center}
\end{figure}


Amino acids can also be plotted although the larger number of amino acids makes comparisons somewhat more difficult (Figure \ref{figAA}). This is somewhat ameliorated by the use of a color scheme based on \href{http://jmol.sourceforge.net/jscolors/}{Jmol} where amino acids with similar characteristics are colored similarly.

<<plotAA, echo=TRUE, eval=FALSE>>=
  fakeAAs<-createFakeAA()
  par(mar=c(3.5,4.4,.5,7),mgp=c(2.5,1,0))
  plotAA(fakeAAs)
@

\begin{figure}[htbp]
\begin{center}
<<showAA, fig=TRUE, height=4.5, width=8, echo=FALSE>>=
<<plotAA>>
@
\caption{An example plot of amino acids}
\label{figAA}
\end{center}
\end{figure}



\end{document}
